File: ios/Freedom/Core/Crypto/PasskeySession.swift
// ═════════════════════════════════════════════════════════════════════════════
//  PasskeySession — PBKDF2 + AES-256-GCM passkey protection
// ═════════════════════════════════════════════════════════════════════════════
//
//  Holds the PBKDF2-derived AES-256 key in memory for the lifetime of the app process.
//
//  - The passkey itself is NEVER stored. Only the PBKDF2 salt + a GCM verifier
//    blob are persisted to Keychain.
//  - The derived key is stored as Data (value type) so it can be
//    zeroed with lock() when the session should end.
//  - encryptField / decryptField protect individual ContactData columns
//    (keys) without encrypting the whole database.
//
// ═════════════════════════════════════════════════════════════════════════════

import Foundation
import CryptoKit
import CommonCrypto
import Security

final class PasskeySession {

    static let shared = PasskeySession()

    private static let VERIFY_PLAIN = "FREEDOM_PASSKEY_V1"

    private static let PBKDF2_ITER: UInt32 = 200_000
    private static let KEY_BITS     = 256
    private static let SALT_BYTES   = 32
    private static let GCM_IV_BYTES = 12

    /// Minimum number of characters required for a passkey.
    static let MIN_PASSKEY_LENGTH = 12

    private var keyBytes: Data?
    private let lock = NSLock()

    // Keychain keys
    private static let keychainService = "com.freedom.passkey"
    private static let keySalt = "passkey_salt"
    private static let keyVerifier = "passkey_verifier"

    var isUnlocked: Bool {
        lock.lock()
        defer { lock.unlock() }
        return keyBytes != nil
    }

    func isPasskeySet() -> Bool {
        return keychainRead(key: Self.keySalt) != nil
    }

    /// First-time setup. Generates a random salt, derives the AES key, stores
    /// salt + verifier, and unlocks the session.
    func setup(passkey: String) {
        var salt = Data(count: Self.SALT_BYTES)
        salt.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, Self.SALT_BYTES, $0.baseAddress!) }

        let derived = derive(passkey: passkey, salt: salt)
        let verifier = aesGcmEncrypt(data: Data(Self.VERIFY_PLAIN.utf8), key: derived)

        keychainWrite(key: Self.keySalt, data: salt)
        keychainWrite(key: Self.keyVerifier, data: verifier)

        lock.lock()
        keyBytes = derived
        lock.unlock()
    }

    /// Unlock on subsequent launches. Returns true if the passkey is correct.
    func unlock(passkey: String) -> Bool {
        guard let salt = keychainRead(key: Self.keySalt),
              let verifier = keychainRead(key: Self.keyVerifier) else {
            return false
        }

        let derived = derive(passkey: passkey, salt: salt)
        do {
            let plain = try aesGcmDecrypt(data: verifier, key: derived)
            if String(data: plain, encoding: .utf8) == Self.VERIFY_PLAIN {
                lock.lock()
                keyBytes = derived
                lock.unlock()
                return true
            } else {
                zeroData(&[derived])
                return false
            }
        } catch {
            var d = derived
            zeroData(&[d])
            return false
        }
    }

    /// Zero and discard the in-memory AES key.
    func lockSession() {
        lock.lock()
        if var k = keyBytes {
            k.resetBytes(in: 0..<k.count)
        }
        keyBytes = nil
        lock.unlock()
    }

    /// Encrypt a field value with the session key.
    /// Empty string is returned unchanged. Returns nil if session is locked.
    func encryptField(_ plaintext: String) -> String? {
        if plaintext.isEmpty { return "" }
        lock.lock()
        guard let k = keyBytes else { lock.unlock(); return nil }
        lock.unlock()
        let encrypted = aesGcmEncrypt(data: Data(plaintext.utf8), key: k)
        return encrypted.base64EncodedString()
    }

    /// Decrypt a field value.
    /// Empty string is returned unchanged. Returns nil if session is locked.
    /// If AES-GCM decryption fails (legacy plaintext), returns encoded unchanged.
    func decryptField(_ encoded: String) -> String? {
        if encoded.isEmpty { return "" }
        lock.lock()
        guard let k = keyBytes else { lock.unlock(); return nil }
        lock.unlock()
        do {
            guard let bytes = Data(base64Encoded: encoded) else { return encoded }
            let plain = try aesGcmDecrypt(data: bytes, key: k)
            return String(data: plain, encoding: .utf8) ?? encoded
        } catch {
            return encoded  // legacy plaintext fallback
        }
    }

    // MARK: - Crypto primitives

    private func derive(passkey: String, salt: Data) -> Data {
        var derivedKey = Data(count: Self.KEY_BITS / 8)
        let passcodeBytes = Array(passkey.utf8)
        derivedKey.withUnsafeMutableBytes { derivedPtr in
            salt.withUnsafeBytes { saltPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passcodeBytes, passcodeBytes.count,
                    saltPtr.baseAddress!.assumingMemoryBound(to: UInt8.self), salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    Self.PBKDF2_ITER,
                    derivedPtr.baseAddress!.assumingMemoryBound(to: UInt8.self),
                    Self.KEY_BITS / 8
                )
            }
        }
        return derivedKey
    }

    private func aesGcmEncrypt(data: Data, key: Data) -> Data {
        let symmetricKey = SymmetricKey(data: key)
        let sealed = try! AES.GCM.seal(data, using: symmetricKey)
        // Format: nonce (12) + ciphertext + tag (16)
        return sealed.nonce.withUnsafeBytes { Data($0) } + sealed.ciphertext + sealed.tag
    }

    private func aesGcmDecrypt(data: Data, key: Data) throws -> Data {
        let iv = data.subdata(in: 0..<Self.GCM_IV_BYTES)
        let rest = data.subdata(in: Self.GCM_IV_BYTES..<data.count)
        let tagStart = rest.count - 16
        let ciphertext = rest.subdata(in: 0..<tagStart)
        let tag = rest.subdata(in: tagStart..<rest.count)

        let nonce = try AES.GCM.Nonce(data: iv)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        let symmetricKey = SymmetricKey(data: key)
        return try AES.GCM.open(sealedBox, using: symmetricKey)
    }

    private func zeroData(_ datas: inout [Data]) {
        for i in 0..<datas.count {
            datas[i].resetBytes(in: 0..<datas[i].count)
        }
    }

    // MARK: - Keychain helpers

    private func keychainWrite(key: String, data: Data) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.keychainService,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
        var add = query
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        SecItemAdd(add as CFDictionary, nil)
    }

    private func keychainRead(key: String) -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.keychainService,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess else { return nil }
        return result as? Data
    }
}

File: main/src/ui/java/freedom/app/security/PasskeySession.kt
package freedom.app.security

import android.content.Context
import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Holds the PBKDF2-derived AES-256 key in memory for the lifetime of the app process.
 *
 * - The passkey itself is NEVER stored. Only the PBKDF2 salt + a GCM verifier
 *   blob are persisted to SharedPreferences.
 * - The derived key is stored as a [ByteArray] (not String) so it can be
 *   zeroed with [lock] when the session should end.
 * - [encryptField] / [decryptField] protect individual ContactData columns
 *   (handshakeKey, otpKey) without encrypting the whole database.
 *
 * Memory safety note:
 *   ByteArray fields can be zeroed; String objects cannot.  Android's
 *   SELinux sandbox prevents other apps from reading this process's heap.
 *   The key lives in memory only while the app process is running; it is
 *   gone when the process is killed.  [lock] zeros it proactively (call
 *   from Activity.onStop if you want to require re-entry on resume).
 */
object PasskeySession {

    private const val PREFS        = "passkey_session"
    private const val KEY_SALT     = "salt"
    private const val KEY_VERIFIER = "verifier"
    private const val VERIFY_PLAIN = "FREEDOM_PASSKEY_V1"

    private const val PBKDF2_ITER  = 200_000
    private const val KEY_BITS     = 256
    private const val SALT_BYTES   = 32
    private const val GCM_IV_BYTES = 12
    private const val GCM_TAG_BITS = 128

    /** Minimum number of characters required for a passkey. */
    const val MIN_PASSKEY_LENGTH = 12

    @Volatile private var keyBytes: ByteArray? = null

    val isUnlocked: Boolean get() = keyBytes != null

    fun isPasskeySet(context: Context): Boolean =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).contains(KEY_SALT)

    /**
     * First-time setup. Generates a random salt, derives the AES key, stores
     * salt + verifier, and unlocks the session.
     * Run on Dispatchers.IO — PBKDF2 at 200k iterations takes ~200–400 ms.
     */
    fun setup(context: Context, passkey: String) {
        val salt     = ByteArray(SALT_BYTES).also { SecureRandom().nextBytes(it) }
        val derived  = derive(passkey, salt)
        val verifier = aesGcmEncrypt(VERIFY_PLAIN.toByteArray(Charsets.UTF_8), derived)
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString(KEY_SALT,     Base64.encodeToString(salt,     Base64.NO_WRAP))
            .putString(KEY_VERIFIER, Base64.encodeToString(verifier, Base64.NO_WRAP))
            .apply()
        keyBytes = derived
    }

    /**
     * Unlock on subsequent launches. Returns true if the passkey is correct.
     * Run on Dispatchers.IO.
     */
    fun unlock(context: Context, passkey: String): Boolean {
        val prefs    = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val saltB64  = prefs.getString(KEY_SALT,     null) ?: return false
        val verB64   = prefs.getString(KEY_VERIFIER, null) ?: return false
        val salt     = Base64.decode(saltB64, Base64.NO_WRAP)
        val verifier = Base64.decode(verB64,  Base64.NO_WRAP)
        val derived  = derive(passkey, salt)
        return try {
            val plain = String(aesGcmDecrypt(verifier, derived), Charsets.UTF_8)
            if (plain == VERIFY_PLAIN) {
                keyBytes = derived
                true
            } else {
                derived.fill(0)
                false
            }
        } catch (_: Exception) {
            derived.fill(0)
            false
        }
    }

    /** Zero and discard the in-memory AES key. */
    fun lock() {
        keyBytes?.fill(0)
        keyBytes = null
    }

    /**
     * Encrypt a field value with the session key.
     * - Empty string is returned unchanged (preserves DB emptiness checks).
     * - Returns null if the session is locked (call [unlock] first).
     */
    fun encryptField(plaintext: String): String? {
        if (plaintext.isEmpty()) return ""
        val k = keyBytes ?: return null
        return Base64.encodeToString(
            aesGcmEncrypt(plaintext.toByteArray(Charsets.UTF_8), k),
            Base64.NO_WRAP
        )
    }

    /**
     * Decrypt a field value.
     * - Empty string is returned unchanged.
     * - Returns null if the session is locked.
     * - If AES-GCM decryption fails (legacy plaintext from before encryption
     *   was introduced), returns [encoded] unchanged so old contacts keep
     *   working until their key is next persisted with encryption.
     */
    fun decryptField(encoded: String): String? {
        if (encoded.isEmpty()) return ""
        val k = keyBytes ?: return null
        return try {
            val bytes = Base64.decode(encoded, Base64.NO_WRAP)
            String(aesGcmDecrypt(bytes, k), Charsets.UTF_8)
        } catch (_: Exception) {
            encoded  // legacy plaintext fallback
        }
    }

    // ── Crypto primitives ─────────────────────────────────────────────────────

    private fun derive(passkey: String, salt: ByteArray): ByteArray {
        val spec    = PBEKeySpec(passkey.toCharArray(), salt, PBKDF2_ITER, KEY_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val bytes   = factory.generateSecret(spec).encoded
        spec.clearPassword()
        return bytes
    }

    private fun aesGcmEncrypt(data: ByteArray, key: ByteArray): ByteArray {
        val iv     = ByteArray(GCM_IV_BYTES).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, iv))
        return iv + cipher.doFinal(data)
    }

    private fun aesGcmDecrypt(data: ByteArray, key: ByteArray): ByteArray {
        val iv         = data.copyOfRange(0, GCM_IV_BYTES)
        val ciphertext = data.copyOfRange(GCM_IV_BYTES, data.size)
        val cipher     = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, iv))
        return cipher.doFinal(ciphertext)
    }
}
